html_text/html_text.py

```diff
@@ -44,30 +54,100 @@ def parse_html(html):
 _whitespace = re.compile(r'\s+')
 _has_trailing_whitespace = re.compile(r'\s$').search
 _has_punct_after = re.compile(r'^[,:;.!?"\)]').search
-_has_punct_before = re.compile(r'\($').search
-
+_has_open_bracket_before = re.compile(r'\($').search
 
-def selector_to_text(sel, guess_punct_space=True):
-    """ Convert a cleaned selector to text.
-    See html_text.extract_text docstring for description of the approach and options.
-    """
-    if guess_punct_space:
 
-        def fragments():
-            prev = None
-            for text in sel.xpath('.//text()').extract():
-                if prev is not None and (_has_trailing_whitespace(prev)
-                                         or (not _has_punct_after(text) and
-                                             not _has_punct_before(prev))):
-                    yield ' '
-                yield text
-                prev = text
+def _normalize_whitespace(text):
+    return _whitespace.sub(' ', text.strip())
 
-        return _whitespace.sub(' ', ''.join(fragments()).strip())
 
+def _html_to_text(tree,
+                  guess_punct_space=True,
+                  guess_layout=False,
+                  newline_tags=NEWLINE_TAGS,
+                  double_newline_tags=DOUBLE_NEWLINE_TAGS):
+    """
+    Convert a cleaned html tree to text.
+    See html_text.extract_text docstring for description of the approach
+    and options.
+    """
+    chunks = []
+
+    class Context:
+        """ workaround for missing `nonlocal` in Python 2 """
+        prev = '\n\n'
+
+    def should_add_space(text, prev):  # type: (str, str) -> bool
+        """ Return True if extra whitespace should be added before text """
+        if prev == '\n' or prev == '\n\n':
+            return False
+        if not _has_trailing_whitespace(prev):
+            if _has_punct_after(text) or _has_open_bracket_before(prev):
+                return False
+        return True
+
+    def get_space_between(text, prev):
+        if not text or not guess_punct_space:
+            return ' '
+        return ' ' if should_add_space(text, prev) else ''
+
+    def add_newlines(tag, context):
+        if not guess_layout:
+            return
+        prev = context.prev
+        if prev == '\n\n':  # don't output more than 1 blank line
+            return
+        if tag in double_newline_tags:
+            context.prev = '\n\n'
+            chunks.append('\n' if prev == '\n' else '\n\n')
+        elif tag in newline_tags:
+            context.prev = '\n'
+            if prev != '\n':
+                chunks.append('\n')
+
+    def add_text(text_content, context):
+        text = _normalize_whitespace(text_content) if text_content else ''
+        if not text:
+            return
+        space = get_space_between(text, context.prev)
+        chunks.extend([space, text])
+        context.prev = text_content
```

Comment:
If `text_content` is `\n\n` or `\n`, then `text` is empty, so function returns earlier. Can't say I understood this before you asked, that's a great question! Probably it makes sense to have the logic more explicit, though I'm not sure how.

Reply:
Thanks for the explanation! Yes, I see that such message won't reach this code part, thanks!

> Probably it makes sense to have the logic more explicit, though I'm not sure how.

I see two ways (not sure how much of an improvement they are):
- have an integer ``prev_newlines`` variable, which can have values 0 (instead of ``prev = text_content``), 1 (instead of ``\n``), and 2 (instead of ``\n\n``)
- have a enum with 3 values

But putting a comment that explains that ``text_content`` here contains some text and can not be equal to newlines is also fine by me.

Reply:
great suggestion @lopuhin, thanks! Did something along these lines here: https://github.com/TeamHG-Memex/html-text/pull/11/commits/7a1b57b9963e349ac1a964af795cc08a0d8aad6c